### PR TITLE
UT for clcreateSubDevices

### PR DESCRIPTION
**Pull request by piotr-wozniak-mobica**

### CODE REVIEW

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3700]: I think this all can be replaced by:

```suggestion
    clGetDeviceInfo_StubWithCallback(clGetDeviceInfo_platform);
    clGetPlatformInfo_StubWithCallback(clGetPlatformInfo_version_1_1);
```

Then, many of the stub functions defined above are no longer needed (all but `clCreateSubDevicesEXT_testDevice_createSubDevices`).

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3710]: These tests that return error codes will all "fail" for the `CL_HPP_ENABLE_EXCEPTIONS` tests, because the error conditions will cause exceptions to be thrown.  Options:

1. They should only be run with `CL_HPP_ENABLE_EXCEPTIONS` is not defined.
2. They should be rewritten so they do not return error codes.

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3682]: Can we test a few more conditions here (and then in the tester below, also)?  Examples:

1. Test that the partition properties are passed through properly.
2. Write output subdevices to the `out_devices` array and verify that they are returned properly.

**Review by bashbaug — COMMENTED** [review]
Here are a few comments and suggested improvements.  The issue where exceptions are being thrown needs to be resolved before merging.

As a meta comment, it would be great to test the core OpenCL 1.2 sub-device functionality in addition to the `cl_ext_device_fission` extension!

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: Changed

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: We need some "fails" to cover all options. CL_HPP_ENABLE_EXCEPTIONS added

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: Changed as suggested

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3641]: This entire function looks like dead code.  If this is not used can it be removed?

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3648]: This is OK for a unit test but do note that `CL_DEVICE_PARTITION_TYPES_EXT` is not a valid partition property according to the cl_ext_device_fission [spec](https://registry.khronos.org/OpenCL/extensions/ext/cl_ext_device_fission.txt).  If we wanted this to be a more real-world test case then we should choose different partition properties, such as perhaps `CL_DEVICE_PARTITION_EQUALLY_EXT` or `CL_DEVICE_PARTITION_BY_AFFINITY_DOMAIN_EXT`.

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3654]: These aren't exactly real-world return values, either:

1. The number of output sub-devices should always be written to if it is non-nullptr, not only if out_devices is nullptr.
2. I don't think it's possible to partition a device into a single sub-device.

For a unit test I suppose this is OK, though it would be better to have more real-world values instead.

File: tests/test_openclhpp.cpp
Comment (bashbaug) [inline tests/test_openclhpp.cpp:3658]: 1. Why does it matter if `nullptr == num`?
2. `CL_DEVICE_NOT_FOUND` is not a valid error code for this function.  Could we pick a valid error code instead, maybe `CL_INVALID_DEVICE` or `CL_INVALID_VALUE`?

**Review by bashbaug — COMMENTED** [review]
Thanks, this is OK as a unit test but I'd prefer to see more real-world parameters and return values if possible.  Can you please take a look at the comments below?

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: Sure, removed.

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: Changed to proper value.

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: Changed to always write number of sub-devices

File: tests/test_openclhpp.cpp
Comment (piotr-wozniak-mobica) [reply tests/test_openclhpp.cpp]: 1. I wanted a function to fail in right place but it can be without this check also
2. Chenged to valid error code

**Review by bashbaug — APPROVED** [review]
There is more that we could do here, but this is an improvement, so no reason to hold off merging it.

**Pull request merged**
